drafts/interesting-bugs-caught-by-no-constant-binary-expression.md

```diff
@@ -0,0 +1,130 @@
+
+### Confused by ?? and || Precedence
+
+When trying to define default values, people get confused with expressions like `a === b ?? c` and assume it will be parsed as `a === (b ?? c)`. When in actuality it will be parsed as `(a === b) ?? c`.
+
+```javascript
+shouldShowWelcome() {
+  return this.viewModel?.welcomeExperience === WelcomeExperience.ForWorkspace ?? true;
+}
+```
+
+_From [VS Code](https://github.com/captbaritone/vscode/blob/ab86e0229d6b4d0cb49cfd6747c92cafcd2bd4af/src/vs/workbench/contrib/testing/browser/testingExplorerView.ts#L118:L118)_
+
+_Aside: Observing how frequently developers get confused by operator precedence inspired me to experiment with [a VS Code extension](https://jordaneldredge.com/blog/a-vs-code-extension-to-combat-js-precedence-confusion/) to visually clarify how precedence gets interpreted._
+
+### Expecting Object to be Compared by Value
+
+Developers coming from other languages where structures are compared by value, rather than by reference, can easily fall into the trap of thinking they can do things like test if an object is empty by comparing with a newly created empty object. Or course in JavaScript, objects are compared by reference, and no value can ever be equal to a newly constructed object literal.
+
+In this example, `hasData` will always be set to true because `data` can never be referentially equal to a newly created object.
+
+```javascript
+hasData = hasData || data !== {};
+```
+
+_From [Firefox](https://hg.mozilla.org/try/rev/0fe5678fb8b71f4eb26f0a153c52d0be45fc5ac1#l3.34) (also: [Firefox](https://hg.mozilla.org/try/rev/0fe5678fb8b71f4eb26f0a153c52d0be45fc5ac1#l1.13))_
+
+### Expecting Empty Objects to be false/null
+
+Another common error class is expecting empty objects to be nullish or falsy. This is likely an easy mistake to make for folks coming from a language like Python where empty lists and dictionaries are falsy.
+
+```javascript
+const newConfigValue = { ...configProfiles } ?? {};
+```
+
+_From [VS Code](https://github.com/captbaritone/vscode/blob/ab86e0229d6b4d0cb49cfd6747c92cafcd2bd4af/src/vs/workbench/contrib/terminal/browser/terminalProfileQuickpick.ts#L126:L126) (also: VS Code [1](https://github.com/captbaritone/vscode/blob/ab86e0229d6b4d0cb49cfd6747c92cafcd2bd4af/src/vs/platform/terminal/node/ptyService.ts#L369:L369), [2](https://github.com/captbaritone/vscode/blob/ab86e0229d6b4d0cb49cfd6747c92cafcd2bd4af/src/vs/workbench/contrib/terminal/browser/terminalProfileResolverService.ts#L484:L484))_
+
+### Is it `>=` or `=>`?
+
+I’ve only seen this particular typo once, but I wanted to include it because think it’s a great example of the unexpected types of bugs this approach can catch.
```

Comment:
```suggestion
I’ve only seen this particular typo once, but I wanted to include it because it’s a great example of the unexpected types of bugs this rule can catch.
```